# Data Analysis Police Dataset Using Pandas

    Title     : Data Analysis Police Dataset Using Pandas  
    Author    : Sopianto Djahar
    Email     : idnsopianto@gmail.com
    Linkedin  : in/sopiantodjahar
    Credit    : https://www.udemy.com/user/saima-aziz/ 

## Analysis Purpose 

* Q1. To remove the column that contains missing values. 
* Q2. For speeding, were men or women stopped more often?
* Q3. Does gender effect who gets sarched during a stop?
* Q4. What is the men stop_duration?
* Q5. To compare the age contributions for each violation.

## Analysis Purpose 

In [151]:
import pandas as pd

## Data Wrangling

In [152]:
# Loading dataset
file_path = "police_data.csv"
police_data = pd.read_csv(file_path)

In [153]:
police_data

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0       1/2/2005      1:55           NaN             M          1985.0   
1      1/18/2005      8:15           NaN             M          1965.0   
2      1/23/2005     23:15           NaN             M          1972.0   
3      2/20/2005     17:15           NaN             M          1986.0   
4      3/14/2005     10:00           NaN             F          1984.0   
...          ...       ...           ...           ...             ...   
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
0            20.0       White                        Speeding   Speeding   
1            40.0       White                        Speeding   Speeding   
2            33.0       White                        Speeding   Speeding   
3            19.0       White                Call for Service      Other   
4            21.0       White                        Speeding   Speeding   
...           ...         ...                             ...        ...   
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 15 columns]

In [154]:
police_data.shape

(65535, 15)

In [155]:
police_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           65535 non-null  object 
 1   stop_time           65535 non-null  object 
 2   country_name        0 non-null      float64
 3   driver_gender       61474 non-null  object 
 4   driver_age_raw      61481 non-null  float64
 5   driver_age          61228 non-null  float64
 6   driver_race         61475 non-null  object 
 7   violation_raw       61475 non-null  object 
 8   violation           61475 non-null  object 
 9   search_conducted    65535 non-null  bool   
 10  search_type         2479 non-null   object 
 11  stop_outcome        61475 non-null  object 
 12  is_arrested         61475 non-null  object 
 13  stop_duration       61475 non-null  object 
 14  drugs_related_stop  65535 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+

## Data Cleaning

### Q1. To remove the column that contains missing values.

In [156]:
# Checking value
police_data.count()

stop_date             65535
stop_time             65535
country_name              0
driver_gender         61474
driver_age_raw        61481
driver_age            61228
driver_race           61475
violation_raw         61475
violation             61475
search_conducted      65535
search_type            2479
stop_outcome          61475
is_arrested           61475
stop_duration         61475
drugs_related_stop    65535
dtype: int64

We have 65535 rows from dataset. If any column which does not have the 65535 rows it means missing value. To check it we will do in the next formula.

In [157]:
# Checking missing value
police_data.isnull()

stop_date  stop_time  country_name  driver_gender  driver_age_raw  \
0          False      False          True          False           False   
1          False      False          True          False           False   
2          False      False          True          False           False   
3          False      False          True          False           False   
4          False      False          True          False           False   
...          ...        ...           ...            ...             ...   
65530      False      False          True          False           False   
65531      False      False          True          False           False   
65532      False      False          True          False           False   
65533      False      False          True           True            True   
65534      False      False          True          False           False   

       driver_age  driver_race  violation_raw  violation  search_conducted  \
0           False        False          False      False             False   
1           False        False          False      False             False   
2           False        False          False      False             False   
3           False        False          False      False             False   
4           False        False          False      False             False   
...           ...          ...            ...        ...               ...   
65530       False        False          False      False             False   
65531       False        False          False      False             False   
65532       False        False          False      False             False   
65533        True         True           True       True             False   
65534       False        False          False      False             False   

       search_type  stop_outcome  is_arrested  stop_duration  \
0             True         False        False          False   
1             True         False        False          False   
2             True         False        False          False   
3             True         False        False          False   
4             True         False        False          False   
...            ...           ...          ...            ...   
65530         True         False        False          False   
65531         True         False        False          False   
65532         True         False        False          False   
65533         True          True         True           True   
65534         True         False        False          False   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 15 columns]

So, here from 65535 rows and 15 columns we found "True" missing value which are "country_name" and "search_type".

In [158]:
# Checking total missing value
police_data.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

To find more detail, we can see the column that has "0" value it means the data is complate.  

In [159]:
# Removing missing value
police_data.drop('country_name', axis =1)

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
3      2/20/2005     17:15             M          1986.0        19.0   
4      3/14/2005     10:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65532  12/6/2012     23:20             M          1985.0        27.0   
65533  12/7/2012      0:23           NaN             NaN         NaN   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65533         NaN                             NaN        NaN   
65534       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 14 columns]

Here, we have just removed "country_name" column from dataset. 

In [160]:
# To verify deleting "country_name" column, we should use this formula to remove permanently.
police_data.drop('country_name', axis =1, inplace = True)

In [161]:
# Removing missing value "search_type"
police_data.drop('search_type', axis =1)

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
3      2/20/2005     17:15             M          1986.0        19.0   
4      3/14/2005     10:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65532  12/6/2012     23:20             M          1985.0        27.0   
65533  12/7/2012      0:23           NaN             NaN         NaN   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65533         NaN                             NaN        NaN   
65534       White                        Speeding   Speeding   

       search_conducted   stop_outcome is_arrested stop_duration  \
0                 False       Citation       False      0-15 Min   
1                 False       Citation       False      0-15 Min   
2                 False       Citation       False      0-15 Min   
3                 False  Arrest Driver        True     16-30 Min   
4                 False       Citation       False      0-15 Min   
...                 ...            ...         ...           ...   
65530             False       Citation       False      0-15 Min   
65531             False        Warning       False      0-15 Min   
65532             False       Citation       False      0-15 Min   
65533             False            NaN         NaN           NaN   
65534             False       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 13 columns]

In [162]:
# To verify deleting "search_type" column, we should use this formula to remove permanently.
police_data.drop('search_type', axis =1, inplace = True)

In [163]:
# Let's check again!
police_data

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
3      2/20/2005     17:15             M          1986.0        19.0   
4      3/14/2005     10:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65532  12/6/2012     23:20             M          1985.0        27.0   
65533  12/7/2012      0:23           NaN             NaN         NaN   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65533         NaN                             NaN        NaN   
65534       White                        Speeding   Speeding   

       search_conducted   stop_outcome is_arrested stop_duration  \
0                 False       Citation       False      0-15 Min   
1                 False       Citation       False      0-15 Min   
2                 False       Citation       False      0-15 Min   
3                 False  Arrest Driver        True     16-30 Min   
4                 False       Citation       False      0-15 Min   
...                 ...            ...         ...           ...   
65530             False       Citation       False      0-15 Min   
65531             False        Warning       False      0-15 Min   
65532             False       Citation       False      0-15 Min   
65533             False            NaN         NaN           NaN   
65534             False       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 13 columns]

### Q2. For speeding, were men or women stopped more often?

In [164]:
police_data.head(5)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted   stop_outcome is_arrested  \
0          Speeding  Speeding             False       Citation       False   
1          Speeding  Speeding             False       Citation       False   
2          Speeding  Speeding             False       Citation       False   
3  Call for Service     Other             False  Arrest Driver        True   
4          Speeding  Speeding             False       Citation       False   

  stop_duration  drugs_related_stop  
0      0-15 Min               False  
1      0-15 Min               False  
2      0-15 Min               False  
3     16-30 Min               False  
4      0-15 Min               False

In [165]:
# Showing all violations "Speeding"
police_data[police_data.violation == 'Speeding']

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
4      3/14/2005     10:00             F          1984.0        21.0   
6       4/1/2005     17:30             M          1969.0        36.0   
...          ...       ...           ...             ...         ...   
65527  12/6/2012     15:26             F          1981.0        31.0   
65529  12/6/2012     16:00             M          1994.0        18.0   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race violation_raw violation  search_conducted stop_outcome  \
0           White      Speeding  Speeding             False     Citation   
1           White      Speeding  Speeding             False     Citation   
2           White      Speeding  Speeding             False     Citation   
4           White      Speeding  Speeding             False     Citation   
6           White      Speeding  Speeding             False     Citation   
...           ...           ...       ...               ...          ...   
65527       White      Speeding  Speeding             False     Citation   
65529       White      Speeding  Speeding             False     Citation   
65530       White      Speeding  Speeding             False     Citation   
65531       White      Speeding  Speeding             False      Warning   
65534       White      Speeding  Speeding             False     Citation   

      is_arrested stop_duration  drugs_related_stop  
0           False      0-15 Min               False  
1           False      0-15 Min               False  
2           False      0-15 Min               False  
4           False      0-15 Min               False  
6           False      0-15 Min               False  
...           ...           ...                 ...  
65527       False      0-15 Min               False  
65529       False      0-15 Min               False  
65530       False      0-15 Min               False  
65531       False      0-15 Min               False  
65534       False      0-15 Min               False  

[37204 rows x 13 columns]

In [166]:
# Checking value_counts that shows each gender
police_data[police_data.violation == 'Speeding'].driver_gender.value_counts()

driver_gender
M    25517
F    11686
Name: count, dtype: int64

### Q3. Does gender effect who gets sarched during a stop?

In [167]:
police_data.head(5)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted   stop_outcome is_arrested  \
0          Speeding  Speeding             False       Citation       False   
1          Speeding  Speeding             False       Citation       False   
2          Speeding  Speeding             False       Citation       False   
3  Call for Service     Other             False  Arrest Driver        True   
4          Speeding  Speeding             False       Citation       False   

  stop_duration  drugs_related_stop  
0      0-15 Min               False  
1      0-15 Min               False  
2      0-15 Min               False  
3     16-30 Min               False  
4      0-15 Min               False

In [168]:
police_data.groupby('driver_gender').search_conducted.sum()

driver_gender
F     366
M    2113
Name: search_conducted, dtype: int64

In [169]:
police_data.search_conducted.value_counts()

search_conducted
False    63056
True      2479
Name: count, dtype: int64

### What is the men stop_duration?

In [170]:
police_data.head(5)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted   stop_outcome is_arrested  \
0          Speeding  Speeding             False       Citation       False   
1          Speeding  Speeding             False       Citation       False   
2          Speeding  Speeding             False       Citation       False   
3  Call for Service     Other             False  Arrest Driver        True   
4          Speeding  Speeding             False       Citation       False   

  stop_duration  drugs_related_stop  
0      0-15 Min               False  
1      0-15 Min               False  
2      0-15 Min               False  
3     16-30 Min               False  
4      0-15 Min               False

In [171]:
# Mapping and datatype casting
police_data.stop_duration.value_counts()

stop_duration
0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: count, dtype: int64

In [172]:
# Changing value in column "stop_duration"
police_data['stop_duration'] = police_data['stop_duration'].map({'0-15 Min': 7.5, '16-30 Min': 23, '30+ Min': 45})

In [173]:
# Checking one more time and make sure "stop_duration" column has replaced by mean!
police_data.head(5)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted   stop_outcome is_arrested  \
0          Speeding  Speeding             False       Citation       False   
1          Speeding  Speeding             False       Citation       False   
2          Speeding  Speeding             False       Citation       False   
3  Call for Service     Other             False  Arrest Driver        True   
4          Speeding  Speeding             False       Citation       False   

   stop_duration  drugs_related_stop  
0            7.5               False  
1            7.5               False  
2            7.5               False  
3           23.0               False  
4            7.5               False

In [174]:
police_data['stop_duration'].mean()

12.001195627419722

### Q5. To compare the age contributions for each violation.

In [175]:
# Describing fuction for statistic measures
police_data.groupby('violation').driver_age.describe()

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0